### PR TITLE
Step12  Implement Sorting by Due Date and Creation Date for Tasks

### PR DESCRIPTION
**Pull request by fanfannb**
### Description:

This PR introduces the functionality to sort tasks by both their creation date and due date, enhancing the task management capabilities of our application.  As part of this update, users can now easily prioritize their tasks based on deadlines and creation times directly from the dashboard.

### Key Changes:

- Database Schema Update:
Added a new column deadline to the tasks table to store the due date of tasks.
- Model Adjustments:
Updated the Task model to include validations ensuring the deadline is optional but, if provided, must be a future date.
- UI Enhancements:
Add a few buttons to the task index page to allow the user to sort the tasks in the following way:
Create_at (ascending and descending)
Deadline (ascending and descending)
- Backend Logic:
Modified the TasksController#index action to handle sorting parameters received from the UI.
Ensured that tasks are fetched with efficient queries to improve performance.

### ステップ12: 終了期限を追加しよう ★ スキップ可能

- タスクに対して、終了期限を登録できるようにしてみましょう
- 一覧画面で、終了期限でソートできるようにしましょう
- specを拡充しましょう
- PRしてレビューをしてもらったら、リリースしてみましょう

截屏2024-06-12 12 12 41

截屏2024-06-12 12 13 03

截屏2024-06-12 18 39 48
截屏2024-06-12 18 39 32
截屏2024-06-12 18 39 14

### CODE REVIEW

**Review by mvincent-yap — CHANGES_REQUESTED** [review]
inconsistent labels:  button show sort by Due date but other show Deadline.

**Comment by mvincent-yap** [comment]
due_date should not be required because older items does not have exisiting due date

**Comment by mvincent-yap** [comment]
also please allow only dates  UI. it is usual practice to only input date instead of datetime

**Comment by mvincent-yap** [comment]
app crashes if you put invalid date like 999999/02/02

File: myapp/app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -3,8 +3,14 @@
...
   before_action :set_task, only: %i[show edit update destroy]
 
+  VALID_SORT_COLUMNS = %w[created_at deadline].freeze
+  VALID_SORT_DIRECTIONS = %w[asc desc].freeze
+
+  # GET /tasks
   def index
-    @tasks = Task.all.order(created_at: :desc)
+    @sort_by = VALID_SORT_COLUMNS.include?(params[:sort_by]) ? params[:sort_by] : 'created_at'
+    @sort_direction = VALID_SORT_DIRECTIONS.include?(params[:sort_direction]) ? params[:sort_direction] : 'asc'
```
Comment (kanipan) [inline myapp/app/controllers/tasks_controller.rb]: Do @sort_by and @sort_direction need to be instance variables?

If not particularly necessary, let's make them local variables (change @sort_by to sort_by). We can make them instance variables when needed.

Other engineers who see instance variables might assume they are used in view files or elsewhere, which could lead to misunderstandings while reading the source code.

Comment (fanfannb) [reply myapp/app/controllers/tasks_controller.rb:12]: Thank you for your comment, but I changed it to a local variable and it won't work.

File: myapp/spec/models/task_spec.rb [hunk 2]
```diff
@@ -4,42 +4,85 @@
...
   end
 
-  it 'is valid without a description' do
-    task = Task.new(title: 'Valid Title', description: '')
+  it 'is valid with an empty description' do
+    task = Task.new(title: 'Valid Title', description: '', deadline: 2.days.from_now)
+    expect(task).to be_valid
+  end
+
+  it 'is not valid without a deadline' do
```
Comment (mvincent-yap) [inline myapp/spec/models/task_spec.rb]: description is incorrect

**Review by mvincent-yap — CHANGES_REQUESTED** [review]

File: myapp/app/controllers/tasks_controller.rb (hunk 1 above)
Comment (fanfannb) [reply myapp/app/controllers/tasks_controller.rb]: Thank you I changed it

**Review by mvincent-yap — APPROVED** [review]
lgtm

**Review by kanipan — APPROVED** [review]

**Comment by kanipan** [comment]
LGTM

寿司ゆき:good_morning

**Pull request merged**